pkg/governance-scripts/contracts/ProtocolFeeControllerMigration.sol

```diff
@@ -0,0 +1,153 @@
+import { IProtocolFeeController } from "@balancer-labs/v3-interfaces/contracts/vault/IProtocolFeeController.sol";
+import { PoolRoleAccounts, PoolConfig } from "@balancer-labs/v3-interfaces/contracts/vault/VaultTypes.sol";
+import { IVaultAdmin } from "@balancer-labs/v3-interfaces/contracts/vault/IVaultAdmin.sol";
+import { IVault } from "@balancer-labs/v3-interfaces/contracts/vault/IVault.sol";
+
+import { ProtocolFeeController } from "@balancer-labs/v3-vault/contracts/ProtocolFeeController.sol";
+import {
+    ReentrancyGuardTransient
+} from "@balancer-labs/v3-solidity-utils/contracts/openzeppelin/ReentrancyGuardTransient.sol";
+
+import { IBasicAuthorizer } from "./IBasicAuthorizer.sol";
+
+/**
+ * @notice Migrate from the original ProtocolFeeController to one with extra events.
+ * @dev These events enable tracking pool protocol fees under all circumstances (in particular, when protocol fees are
+ * initially turned off). It also adds some infrastructure that makes future migrations easier, and removes redundant
+ * poolCreator storage.
+ *
+ * This simple migration assumes:
+ * 1) There are no pools with pool creators
+ * 2) There are no pools with protocol fee exemptions or overrides
+ * 3) Migrating the complete list of pools can be done in a single transaction.
+ *
+ * These simplifications enable simply calling `migrateFeeController` once with the complete list of pools.
+ *
+ * After the migration, the Vault will point to the new fee controller, and any collection thereafter will go there.
+ * If there are any residual fee amounts in the old fee controller (i.e., that were collected but not withdrawn),
+ * governance will still need to withdraw from the old fee controller. Otherwise, no further interaction with the old
+ * controller is necessary.
+ *
+ * Associated with `20250221-protocol-fee-controller-migration`.
+ */
+contract ProtocolFeeControllerMigration is ReentrancyGuardTransient {
+    IProtocolFeeController public immutable oldFeeController;
+    IProtocolFeeController public immutable newFeeController;
+
+    IVault public immutable vault;
+
+    // IAuthorizer with interface for granting/revoking roles.
+    IBasicAuthorizer internal immutable _authorizer;
+
+    // Set when the operation is complete and all permissions have been renounced.
+    bool internal _finalized;
+
+    /**
+     * @notice Attempt to deploy this contract with invalid parameters.
+     * @dev ProtocolFeeController contracts return the address of the Vault they were deployed with. Ensure that both
+     * the old and new controllers reference the same vault.
+     */
+    error InvalidFeeController();
+
+    /// @notice Migration can only be performed once.
+    error AlreadyMigrated();
+
+    constructor(IVault _vault, IProtocolFeeController _newFeeController) {
+        oldFeeController = _vault.getProtocolFeeController();
+
+        // Ensure valid fee controllers. Also ensure that we are not trying to operate on the current fee controller.
```

Comment:
Could avoid all this by simply deploying a new ProtocolFeeController here (the only argument is the Vault), but we don't usually do that in migrations. Would be a simplification, though.